mon: add nvmeof healthchecks

VallariAg opened the pull request:
Add NVMeofGwMap::get_health_checks which raises
NVMEOF_SINGLE_GATEWAY if any of the groups have
1 gateway.

In NVMeofGwMon, call `encode_health` and `load_health` 
to register healthchecks. This will add nvmeof healthchecks 
to "ceph health" output.

In src/mon/NVMeofGwMap.cc,
add warning NVMEOF_GATEWAY_DOWN when any
gateway is in GW_UNAVAILABLE state.

Fixes (partially): https://github.com/ceph/ceph-nvmeof/issues/565


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

VallariAg committed:
mon: add nvmeof healthchecks

Add NVMeofGwMap::get_health_checks which raises
NVMEOF_SINGLE_GATEWAY if any of the groups have
1 gateway.

In NVMeofGwMon, call `encode_health` and `load_health`
to register healthchecks. This will add nvmeof healthchecks
to "ceph health" output.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

pcuzner commented on src/mon/NVMeofGwMap.cc:
```diff
@@ -873,6 +875,29 @@ struct CMonRequestProposal : public Context {
   }
 };
 
+void NVMeofGwMap::get_health_checks(health_check_map_t *checks) const 
```
since you're looking at the membership of a gateway group, would it make sense to also add a healthcheck for when a gateway is down/inaccessible here - that's the scenario that will hurt the most, IMO

VallariAg committed:
mon: add warning NVMEOF_GATEWAY_DOWN

In src/mon/NVMeofGwMap.cc,
add warning NVMEOF_GATEWAY_DOWN when any
gateway is in GW_UNAVAILABLE state.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg commented:
Results: https://pulpito.ceph.com/vallariag-2024-10-03_11:37:32-nvmeof:thrash-main-distro-default-smithi/
(using [this](https://shaman.ceph.com/builds/ceph/wip-nvmeof-healthcheck-vallari-centos9-only/84eba01295c143c71577661d19fa38c418514e84/default/411932/) build)

When 2 of total 4 gateways are down:
```
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:  cluster:
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    id:     48fd6298-817e-11ef-bafb-efdc52797490
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    health: HEALTH_WARN
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:            2 gateway(s) are in unavailable state; gateway might be down, try to redeploy.
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:  services:
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    mon:    3 daemons, quorum a,c,b (age 12m)
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    mgr:    x(active, since 13m)
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    osd:    8 osds: 8 up (since 8m), 8 in (since 9m)
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:    nvmeof: 2 gateways active (2 hosts)
2024-10-03T12:10:32.773 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:  data:
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:    pools:   2 pools, 33 pgs
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:    objects: 486 objects, 1020 MiB
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:    usage:   2.1 GiB used, 713 GiB / 715 GiB avail
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:    pgs:     33 active+clean
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:  io:
2024-10-03T12:10:32.774 INFO:teuthology.orchestra.run.smithi094.stdout:    client:   8.4 MiB/s rd, 9.1 MiB/s wr, 312 op/s rd, 323 op/s wr
```
Then when they're all back up: 
```
2024-10-03T12:11:50.736 INFO:teuthology.orchestra.run.smithi094.stdout:HEALTH_OK
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:  cluster:
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:    id:     48fd6298-817e-11ef-bafb-efdc52797490
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:    health: HEALTH_OK
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:  services:
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:    mon:    3 daemons, quorum a,c,b (age 13m)
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:    mgr:    x(active, since 14m)
2024-10-03T12:11:51.086 INFO:teuthology.orchestra.run.smithi094.stdout:    osd:    8 osds: 8 up (since 10m), 8 in (since 10m)
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    nvmeof: 4 gateways active (4 hosts)
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:  data:
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    pools:   2 pools, 33 pgs
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    objects: 486 objects, 1024 MiB
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    usage:   2.9 GiB used, 712 GiB / 715 GiB avail
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    pgs:     33 active+clean
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:  io:
2024-10-03T12:11:51.087 INFO:teuthology.orchestra.run.smithi094.stdout:    client:   7.6 MiB/s rd, 8.1 MiB/s wr, 308 op/s rd, 318 op/s wr
```

baum reviewed (approved):
lgtm 🖖

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -1633,6 +1633,25 @@ We encourage you to fix this by making the weights even on both dividing buckets
+NVMEOF_SINGLE_GATEWAY
+_____________________
+
+One of the gateway groups have only one gateway. This is not ideal because it makes
```
s/have/has/

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -1633,6 +1633,25 @@ We encourage you to fix this by making the weights even on both dividing buckets
+high availability (HA) impossible with a single gatway in a group. This can lead to 
+problems with failover and failback operations for the NVMe-oF gateway.
+
+It's recommended to have multiple nvmeof gateways in a group.
```
Suggest `NVMeoF` consistently

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -1633,6 +1633,25 @@ We encourage you to fix this by making the weights even on both dividing buckets
+NVMEOF_GATEWAY_DOWN
+___________________
+
+Some of the gateways are in GW_UNAVAILABLE state. If a nvmeof daemon has crashed, 
```
in the

VallariAg commented:
@anthonyeleven thank you for reviewing, I've updated the docs!

VallariAg committed:
qa/suites/nvmeof: add nvmeof warnings to log-ignorelist

Add NVMEOF_SINGLE_GATEWAY and NVMEOF_GATEWAY_DOWN
warnings to nvmeof:thrash job's log-ignorelist

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

athanatos commented:
jenkins test api

VallariAg commented:
jenkins test api

VallariAg commented:
jenkins test make check

zdover23 commented:
The documentation changes that have been made in this document have been backported here: https://github.com/ceph/ceph/pull/60930#issuecomment-2517569596

These changes impact ``doc/rados/operations/health-checks.rst``.
